plugin/pki/backend_test.go

```diff
@@ -1,596 +1,53 @@
-	data.signCSR = true
-	data.provider = "tpp"
-
-	var err error
-	//Generating CSR for test
-	certificateRequest := x509.CertificateRequest{}
-	certificateRequest.Subject.CommonName = data.cn
-	certificateRequest.DNSNames = append(certificateRequest.DNSNames, data.dns_ns, data.dns_ip)
-	certificateRequest.IPAddresses = []net.IP{net.ParseIP(data.dns_ip)}
-	org := os.Getenv("CERT_O")
-	if org != "" {
-		certificateRequest.Subject.Organization = append(certificateRequest.Subject.Organization, org)
-	}
-
-	//Generating pk for test
-	priv, err := rsa.GenerateKey(r.Reader, 2048)
-	if err != nil {
-		t.Fatal(err)
-	}
-	data.csrPK = pem.EncodeToMemory(
-		&pem.Block{
-			Type:  "RSA PRIVATE KEY",
-			Bytes: x509.MarshalPKCS1PrivateKey(priv),
-		},
-	)
-
-	csr, err := x509.CreateCertificateRequest(r.Reader, &certificateRequest, priv)
-	if err != nil {
-		csr = nil
-	}
-	pemCSR := strings.TrimSpace(string(pem.EncodeToMemory(&pem.Block{
-		Type:  "CERTIFICATE REQUEST",
-		Bytes: csr,
-	})))
-
-	coreConfig := &vault.CoreConfig{
-		LogicalBackends: map[string]logical.Factory{
-			"pki": Factory,
-		},
-		Logger: hclog.NewNullLogger(),
-	}
-	cluster := vault.NewTestCluster(t, coreConfig, &vault.TestClusterOptions{
-		HandlerFunc: vaulthttp.Handler,
-	})
-	cluster.Start()
-	defer cluster.Cleanup()
-
-	client := cluster.Cores[0].Client
-	err = client.Sys().Mount("pki", &api.MountInput{
-		Type: "pki",
-		Config: api.MountConfigInput{
-			DefaultLeaseTTL: "16h",
-			MaxLeaseTTL:     "32h",
-		},
-	})
+//Testing all endpoints with fake vcert CA
+func TestEndpoints(t *testing.T) {
+	integrationTestEnv, err := newIntegrationTestEnv()
 	if err != nil {
 		t.Fatal(err)
 	}
 
-	_, err = client.Logical().Write("pki/roles/example", map[string]interface{}{
-		"generate_lease":    true,
-		"tpp_url":           os.Getenv("TPPURL"),
-		"tpp_user":          os.Getenv("TPPUSER"),
-		"tpp_password":      os.Getenv("TPPPASSWORD"),
-		"zone":              os.Getenv("TPPZONE"),
-		"trust_bundle_file": os.Getenv("TRUST_BUNDLE"),
-		//"service_generated_cert": true,
-	})
-	if err != nil {
-		t.Fatal(err)
-	}
-
-	resp, err := client.Logical().Write("pki/sign/example", map[string]interface{}{
-		"csr": pemCSR,
-	})
-	if err != nil {
-		t.Fatal(err)
-	}
-
-	if resp.Data["certificate"] == "" {
-		t.Fatalf("expected a cert to be generated")
-	}
-
-	data.cert = resp.Data["certificate"].(string)
-
-	checkStandartCert(t, data)
+	t.Run("fake create role", integrationTestEnv.FakeCreateRole)
+	t.Run("fake list roles", integrationTestEnv.FakeListRole)
+	t.Run("fake issue", integrationTestEnv.FakeIssueCertificate)
+	t.Run("fake sign", integrationTestEnv.FakeSignCertificate)
+	t.Run("fake list certificates", integrationTestEnv.FakeListCertificate)
+	t.Run("fake read certificate by cn", integrationTestEnv.FakeReadCertificateByCN)
+	t.Run("fake read certificate by serial", integrationTestEnv.FakeReadCertificateBySerial)
+	t.Run("fake revoke certificate", integrationTestEnv.FakeRevokeCertificate)
```

Comment:
I'm curious about why revocation isn't implemented yet? Should the endpoint still be included if it's not implemented?

Reply:
We're planning to add revocation in future but for now didn't have time to implement it. You think we should remove it?

Reply:
@tyrannosaurus-becks we think the "not implemented" error is appropriate since it is a feature that the Vault native PKI supports and we don't have parity.

Reply:
@tyrannosaurus-becks we didn't implement revocation simply because it isn't a priority for the types of use cases we targeted (i.e. DevOps), same opinions that are expressed around revocation in the [Vault docs](https://www.vaultproject.io/docs/secrets/pki/index.html).  I anticipate that we will add it in the future and, as @arykalin said, feel it is appropriate for "not implemented" error to be returned since POST /pki/revoke is native Vault functionality.

Reply:
I see what you're saying. I'm not aware of it being a paradigm followed in Vault. I don't see any harm in keeping it though.